### PR TITLE
Refactoring 2-structured-data tests to use py.test

### PR DESCRIPTION
**Pull request by jonparrott**
@waprin Note that this is being merged into the `pytest-refactor` branch, which we'll merge into master when all is done.

Also, question for you: I can put all of the test functions in a class and mark the whole class as flaky. That'll keep me from having to mark each test case as flaky. Your call.

### CODE REVIEW

File: 2-structured-data/tests/test_crud.py [hunk 1]
```diff
@@ -12,138 +12,115 @@
...
-        with self.app.app_context():
-            self.model = bookshelf.get_model()
-
-        self.ids_to_delete = []
-
-        # Monkey-patch create so we can track the IDs of every item
-        # created and delete them during tearDown.
-        self.original_create = self.model.create
+@py.test.yield_fixture(params=['datastore', 'cloudsql', 'mongodb'])
+def app(request, monkeypatch):
```
Comment (jerjou) [inline 2-structured-data/tests/test_crud.py]: monkeypatch here looks vestigial

Comment (jonparrott) [inline 2-structured-data/tests/test_crud.py]: That's a... shockingly accurate way of putting it.

File: 2-structured-data/tests/test_crud.py [hunk 2]
```diff
@@ -12,138 +12,115 @@
...
-        self.createBooks(11)
+    monkeypatch.setattr(model, 'create', tracking_create)
 
-        with self.app.test_client() as c:
-            rv = c.get('/books/')
+    yield model
 
-        assert rv.status == '200 OK'
+    # Delete all items that we created during tests.
+    list(map(model.delete, ids_to_delete))
```
Comment (jerjou) [inline 2-structured-data/tests/test_crud.py]: Shouldn't need to cast to list here, no?

Comment (jonparrott) [inline 2-structured-data/tests/test_crud.py]: in Python 3, map is an iterator. list forces the iterator to drain.

File: 2-structured-data/tests/test_crud.py [hunk 3]
```diff
@@ -12,138 +12,115 @@
...
-            self.ids_to_delete.append(res['id'])
-            return res
+    with app.test_request_context():
+        yield app
 
-        self.model.create = tracking_create
 
-    def tearDown(self):
+@py.test.yield_fixture
+def model(monkeypatch, app):
```
Comment (jerjou) [inline 2-structured-data/tests/test_crud.py]: Can you add a comment that including 'app' here puts the model in the test_request_context? Seems like that's important, and I can see myself coming back here later and being like "`app` is unused! DELETE"

Comment (jonparrott) [inline 2-structured-data/tests/test_crud.py]: Will do.

File: 2-structured-data/tests/test_crud.py (hunk 2 above)
Comment (jerjou) [inline 2-structured-data/tests/test_crud.py]: Yeeeeeaaaaah.... I should migrate my brain from python 2 to 3 one of these days...

**Comment by jerjou** [comment]
LGTM
travis hates it though.

**Comment by jonparrott** [comment]
Travis didn't fail on this particular folder, so I'm ignoring it.

Will wait for @waprin to get a chance to review.

**Comment by jonparrott** [comment]
@jerjou @waprin I could structure the tests like this:

```python

@py.test.yield_fixture(params=['datastore', 'cloudsql', 'mongodb'])
def app(request):
    ...


@py.test.yield_fixture
def model(monkeypatch, app):
    ...


@flaky(rerun_filter=flaky_filter)
@py.test.mark.usefixtures('app', 'model')
class TestCrudOperations(object):
    def test_list(self, app, model):
        ...

    def test_add(app):
        ...

    def test_edit(app, model):
        ...

    def test_delete(app, model):
        ...

```

This would require only one use of `flaky`, and would remove the unused `model` fixed from `test_add` while still making sure the fixture is used to clean up any created resources. WDYT?

**Comment by jerjou** [comment]
Either way - as long as there's a comment alluding to the dependency used for cleanup, so future-self won't go on a refactoring spree that messes things up mysteriously :)

File: 2-structured-data/tests/test_crud.py (hunk 3 above)
Comment (waprin) [inline 2-structured-data/tests/test_crud.py]: Overall pytest looks neat, but the whole fixture injection is pretty confusing. I definitely think we need some doc strings explaining where monkeypatch and app are coming from. I still have no idea where monkeypatch is coming from.

Comment (waprin) [inline 2-structured-data/tests/test_crud.py]: Ok monkeytest is a builtin fixture I guess. Yeah a note on that would be helpful.

**Comment by waprin** [comment]
These tests are awesome, I would just like to see the pytest magic explained.

**Comment by jonparrott** [comment]
> These tests are awesome, I would just like to see the pytest magic explained.

I think the class-layout will help with that (`usesfixtures` helps), but I do want to avoid teaching py.test or talking down to the developer.

Let me switch to that, I think it's all-around easier to grok.

**Comment by jonparrott** [comment]
@waprin @jerjou re-organized and added lots of comments and docstrings.

**Comment by waprin** [comment]
lgtm

**Comment by jonparrott** [comment]
Woohoo.

**Pull request merged**